### PR TITLE
Improve error handling when converting classic to block menus

### PR DESCRIPTION
**Pull request by dlh01**
Trac ticket: https://core.trac.wordpress.org/ticket/58823

### CODE REVIEW

**Comment by getdave** [comment]
Thank you for this PR and catching those Issues. Looks like the unit tests will need updating to conform to the changes. Do you think you will be able (have capacity) to do that?

Also cc'ing @spacedmonkey as he may need to oversee this being merged into Core.

**Comment by getdave** [comment]
This will also resolve https://github.com/WordPress/gutenberg/issues/52481

**Comment by dlh01** [comment]
Yep, it's on my list to update in the next week or so, I hope.

**Comment by dlh01** [comment]
@getdave Test updated!

**Review by anton-vlasenko — APPROVED** [review]
The changes proposed in this PR fix discrepancies in how errors are handled in different situations.
LGTM (but please see https://github.com/WordPress/wordpress-develop/pull/4858#pullrequestreview-1587133320).

File: tests/phpunit/tests/editor/classic-to-block-menu-converter.php [hunk 1]
```diff
@@ -214,7 +214,7 @@ public function test_returns_empty_array_for_menus_with_no_items() {
 
 		$this->assertEmpty( $blocks, 'Result should be empty.' );
 
-		$this->assertIsArray( $blocks, 'Result should be empty array.' );
+		$this->assertIsString( $blocks, 'Result should be empty string.' );
```
Comment (anton-vlasenko) [inline tests/phpunit/tests/editor/classic-to-block-menu-converter.php]:
```suggestion
		$this->assertSame( '', $blocks, 'Result should be empty string.' );
```

**Review by anton-vlasenko — COMMENTED** [review]
I think this should be changed, @dlh01.

File: tests/phpunit/tests/editor/classic-to-block-menu-converter.php (hunk 1 above)
Comment (dlh01) [reply tests/phpunit/tests/editor/classic-to-block-menu-converter.php]: Done in [0f530d1](https://github.com/WordPress/wordpress-develop/pull/4858/commits/0f530d1e039a15973cce67af4d2163002524303d).

File: src/wp-includes/class-wp-classic-to-block-menu-converter.php [hunk 2]
```diff
@@ -20,7 +20,7 @@ class WP_Classic_To_Block_Menu_Converter {
 	 * @since 6.3.0
 	 *
 	 * @param WP_Term $menu The Menu term object of the menu to convert.
-	 * @return string the serialized and normalized parsed blocks.
+	 * @return string|WP_Error The serialized and normalized parsed blocks or a WP_Error object.
```
Comment (hellofromtonya) [inline src/wp-includes/class-wp-classic-to-block-menu-converter.php]: For consistency and completeness, I'd suggest the following:

```suggestion
	 * @return string|WP_Error The serialized and normalized parsed blocks on success,
	 *                         an empty string when there are no menus to convert,
	 *                         or WP_Error on invalid menu.
```

File: src/wp-includes/class-wp-classic-to-block-menu-converter.php [hunk 3]
```diff
@@ -34,7 +34,7 @@ public static function convert( $menu ) {
 		$menu_items = wp_get_nav_menu_items( $menu->term_id, array( 'update_post_term_cache' => false ) );
 
 		if ( empty( $menu_items ) ) {
-			return array();
+			return '';
```
Comment (hellofromtonya) [inline src/wp-includes/class-wp-classic-to-block-menu-converter.php:39]: Is this a BC (backward compat) issue?

tl;dr I think no.

Why?

The method shipped in 6.3.0 is clearly documented as returning a `string`, not an `array`. Returning an `array` then is a bug that needs to be fixed.

**Review by hellofromtonya — APPROVED** [review]
LGTM 👍 Should be commit ready

**Comment by hellofromtonya** [comment]
Committed via https://core.trac.wordpress.org/changeset/56422. Thank you everyone for your contributions!
